File: src/main/scala/com/mogobiz/sample/handlers/MyCouponHandler.scala
package com.mogobiz.sample.handlers

import com.mogobiz.run.handlers.CouponHandler
import com.mogobiz.run.model.Mogobiz.{ReductionRule, ReductionRuleType}
import com.mogobiz.run.model.StoreCartItemWithPrice


class MyCouponHandler extends CouponHandler {
  override protected def applyReductionRule(rule: ReductionRule, cartItem: StoreCartItemWithPrice, quantity: Long, price: Long, totalPrice: Long, cartItems: List[StoreCartItemWithPrice]): Long = {
    rule.xtype match {
      case ReductionRuleType.CUSTOM if rule.discount.forall(_.startsWith("MyCustomRule")) && cartItems.size > 10 =>
        computeDiscount(Some("10%"), totalPrice)
      case _ => super.applyReductionRule(rule, cartItem, quantity, price, totalPrice, cartItems)
    }
  }
}

File: src/main/java/com/mogobiz/samplej7/handlers/MyCouponHandler.java
package com.mogobiz.samplej7.handlers;

import com.mogobiz.run.handlers.CouponHandler;
import com.mogobiz.run.model.Mogobiz.ReductionRuleType;
import com.mogobiz.run.model.Mogobiz.ReductionRule;
import com.mogobiz.run.model.StoreCartItemWithPrice;
import scala.Function1;
import scala.Some;
import scala.collection.immutable.List;

public class MyCouponHandler extends CouponHandler {
  @Override
  public long applyReductionRule(ReductionRule rule, StoreCartItemWithPrice cartItem, long quantity, long price, long totalPrice, List<StoreCartItemWithPrice> cartItems) {
    if (rule.xtype().equals(ReductionRuleType.CUSTOM()) &&
        rule.discount().forall(new Function1<String, Object>() {
          @Override
          public Object apply(String v) {
            return v.startsWith("MyCustomRule");
          }
        }) &&
        cartItems.size() > 10) {
      return computeDiscount(Some.apply("10%"), totalPrice);
    } else {
      return super.applyReductionRule(rule, cartItem, quantity, price, totalPrice, cartItems);
    }
  }
}
